### DISCUSSION TITLE
How to debug a tock application over JTAG?

### DISCUSSION THREAD

**Discussion by vChavezB** [tock/tock#3371] [Q&A] [answered]
I am a newcomer to Tock and would like to know if there is documentation on how to debug apps.

I saw [this ](https://github.com/tock/tock/tree/master/doc/debugging) about debugging but is this only to debug the kernel?

I have worked with dynamic loaders for ARM MCUs and from what I understand is that you need the DWARF debug section from the ELF file of the "app". My solution to debug Position independent code has been to use a custom python gdb script that gets metadata from the kernel and then checks if any app is related to the Elf file with debug symbols I want to load (automatically via python gdb). Afterwards I get the memory in Flash/RAM where my app is installed and use  `add-symbol-file my_app.elf APP_INSTALLATION_ADDR`

Im curious to know if this process similar for Tock and if there is any documentation on how to debug apps (with a debugger) ?

**Comment by hudson-ayers** [comment, marked as answer]
There are instructions in the board-specific README's for several boards some of which is pretty generally applicable:

https://github.com/tock/tock/tree/master/boards/imix#debugging

https://github.com/tock/tock/tree/master/boards/nordic/nrf52dk#debugging

https://github.com/tock/tock/tree/master/boards/opentitan#qemu-gdb-debugging-earlgrey-cw310

None of those talk much about debugging apps specifically though, they all focus on kernel debugging. My experience has been that using add-symbol-file to add debug info has been sufficient for RISC-V apps, since they are fixed-position by default. The PIC approach you describe sounds like it should generally work fine for ARM apps

**Reply by vChavezB** [reply]
Yes the approach I mentioned was something I used for a project where we needed to debug PIC binaries so we made a script to automate this.

I will have a look on the links thanks.

**Reply by potto216** [reply]
@vChavezB were you successful debugging an app? I want to be able to debug libtock-rs apps and any advice is appreciated.

**Reply by hudson-ayers** [reply]
all libtock-rs apps are fixed position, just using add_symbol_file your_app.elf may well be sufficient to get basic app debugging working if you are already at the point where you have kernel debugging working

**Reply by potto216** [reply]
Thanks @hudson-ayers I'll try that out. I'll first go through the kernel debugging instructions and then try that with my app
